ouroboros-consensus/test-storage/Test/Ouroboros/StateMachine.hs

```diff
@@ -0,0 +1,462 @@
+    | Close                       Bool (Fd, (Reference (Opaque (FsHandle IOFSE)) r))
+    | Seek                        Bool (Fd, (Reference (Opaque (FsHandle IOFSE)) r)) SeekMode Word64
+    | Get                         Bool (Fd, (Reference (Opaque (FsHandle IOFSE)) r)) Int
+    | Put                         Bool (Fd, (Reference (Opaque (FsHandle IOFSE)) r)) BL.ByteString
+--    | Truncate                    Bool (Fd, (Reference (Opaque (FsHandle IOFSE)) r)) Word64
+    | CreateDirectory             Bool FsPath
+    | CreateDirectoryIfMissing    Bool Bool FsPath
+    | ListDirectory               Bool FsPath
+    | DoesDirectoryExist          Bool FsPath
+    | DoesFileExist               Bool FsPath
+      deriving (Generic1, Rank2.Foldable, Rank2.Traversable, Rank2.Functor)
+
+deriving instance Show (Command Symbolic)
+deriving instance Show (Command Concrete)
+
+data Response (r :: Type -> Type)
+  = OpenR                            (Either FsError (Reference (Opaque (FsHandle IOFSE)) r))
+  | CloseR                           (Either FsError ())
+  | SeekR                            (Either FsError Word64)
+  | GetR                             (Either FsError ByteString)
+  | PutR                             (Either FsError Word64)
+--  | TruncateR                        (Either FsError ())
+  | CreateDirectoryR                 (Either FsError ())
+  | CreateDirectoryIfMissingR        (Either FsError ())
+  | ListDirectoryR                   (Either FsError [String])
+  | DoesDirectoryExistR              (Either FsError Bool)
+  | DoesFileExistR                   (Either FsError Bool)
+  deriving (Generic1, Rank2.Foldable)
+
+deriving instance Show (Response Symbolic)
+deriving instance Show (Response Concrete)
+
+initModel :: Model r
+initModel = Model . Opaque $ (newEmptyMockFS, 0, 0, Map.empty)
+
+preconditions :: Model Symbolic -> Command Symbolic -> Logic
+preconditions m@(Model (Opaque (fs, n, nextFd, mp))) cmd = case cmd of
+    Get _ (fd, _) _ ->
+        let (mockHandle, _) = mp Map.! fd
+            ioMode = mockHandleIOMode2 mockHandle
+        -- this condition helps avoiding EBADF errors by trying to read from invalid fds.
+        -- TODO: we could reify this error is FsError and delete this precondition.
+        in Boolean $ ioMode `Prelude.elem` [ReadMode, ReadWriteMode]
+    Put _ (fd, _) _ ->
+        let (mockHandle, _) = mp Map.! fd
+            ioMode = mockHandleIOMode2 mockHandle
+        -- this condition helps avoiding EBADF errors by trying to wrote to invalid fds.
+        -- TODO: we could reify this error is FsError and delete this precondition.
+        in Boolean $ ioMode `Prelude.elem` [WriteMode, ReadWriteMode]
+    _ -> Top
+
+standardPostCondition :: Either FsError a -> Either FsError b -> Logic
+standardPostCondition (Left e1) (Left e2) =
+    if e1 `sameFsError` e2 then Top
+    else error $ "Sim returned error " ++ show e1 ++ " while IO returned " ++ show e2
```

Comment:
Instead of throwing an error you can also return false with an annotation, i.e.: `Bot .// "Sim returned ..."`.

Reply:
That's great thanks!